### PR TITLE
Add directory paging for rendering code trees

### PR DESCRIPTION
**Pull request by zeripath**
This PR simply adds the ability to make directory rendering be paged.

Coupled with the previous improvements to last commit info generation
this should speed up directory rendering for large repositories.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: modules/setting/setting.go [hunk 1]
```diff
@@ -248,6 +249,7 @@ var (
 		IssuePagingNum:      10,
 		RepoSearchPagingNum: 10,
 		MembersPagingNum:    20,
+		DirectoryPagingNum:  50,
```
Comment (lafriks) [inline modules/setting/setting.go:252]: Default should be larger imho, I would say 100

Comment (zeripath) [reply modules/setting/setting.go:252]: Have you tried this?

I thought 100 initially but actually 50 was long enough IMO. 

There are also a few performance arguments to be made for using a number <70.

Comment (lafriks) [reply modules/setting/setting.go:252]: 50 will make paging even for gitea repo root directory listing...

**Comment by lafriks** [comment]
Imho we should reconsider take on this, we shoud make entry list commit info as async instead of paging, or at least make paging larger like at least 100 + async commit info

**Comment by zeripath** [comment]
> Imho we should reconsider take on this, we shoud make entry list commit info as async instead of paging, or at least make paging larger like at least 100 + async commit info

You mean #16467?

---

The directories need to paged anyway. Consider this tree:

https://cgit.freebsd.org/ports/tree/devel

It's insane and #16467 is slightly pathological for it because the javascript looking for the missing commits is quite slow due to the insane number of items it needs to generate commit information for.

> 50 will make paging even for gitea repo root directory listing...

Is that a bad idea? 63 items is kinda long to scroll past to get to the README.

---

I chose 50 because there is a genuine performance improvement below 70 items and the UI doesn't feel too restrictive. Generating last commit info for more than 70 means walking and generating all the last commit information (until you get below 70 items) because git's internal algorithm for shortcutting commits starts to fail at that point.  That means that on large repositories every page would get the same initial delay.

(The 70 number comes from: 

https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L32

Feel free to play with that number there and a copy of git.git and ports.git. )

If you really want this to be 100 instead of 50 then to get a performance improvement from it we're going to have to change the algorithm to cache everything it sees up to the point it can start shortcutting. Otherwise this would actually represent an increase in load and likely a worsening of performance.

---

Changing the algorithm to cache everything it is seeing up to the point that it can start shortcutting whilst possible would require passing in two sets of information in to WalkGitLog. 

https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L270

The most performant solution would be to:

* As per #16467 we'd need to pass in the cache into this function.
* IFF we have a lastcommitcache & the number of interested paths > 70
* Either: pass in all the paths in the treepath and a []bool slice marking them as interest with a total count OR: Generate these in WalkGitLog. (Clearly the second is quite inefficient but it's keeping the interface a bit cleaner.) Generating the data itself would be done around: https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L281
* Create allPath2idx, a allMaxpathlen and a allIdx2idx []int: https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L300-L301
* Create allChanged []bool: https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L325
* As we walk the commits we'd have to pass in the full path2idx and full changed - then map those on the initial map - likely we'd have to pass allPath2idx allIdx2Idx and allChanged https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L334
caching results as we find them (as per #16467) and add additional cleaning check for original path2idx etc here:
https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L252-L260 
* Then once we pass the check in https://github.com/go-gitea/gitea/blob/e76f8cac9a2ba727bec6b5beab2496be9dafabef/modules/git/log_name_status.go#L370 and can restart the log walker we can just replace the allPath2Idx set with the remaining of interest set and stop the double checking.

File: routers/web/repo/view.go [hunk 2]
```diff
@@ -232,6 +222,42 @@ func renderDirectory(ctx *context.Context, treeLink string) {
...
+		page = 1
+	}
+	pageSize := setting.UI.DirectoryPagingNum
+
+	if pageSize > 1 {
+		pager := context.NewPagination(len(entries), pageSize, page, 5)
+
+		ctx.Data["Page"] = pager
+
+		first := pageSize * (page - 1)
```
Comment (6543) [inline routers/web/repo/view.go]: why not use modules/util/PaginateSlice() ?

**Comment by 6543** [comment]
I like the idear but it does not speed up the view - somehow each page does refill commit cache on first hit

File: routers/web/repo/view.go (hunk 2 above)
Comment (zeripath) [reply routers/web/repo/view.go]: I wasn't aware of it. I'll make a few minor improvements. (Repeatedly calling Slice is not a good idea.)

**Comment by zeripath** [comment]
> I like the idear but it does not speed up the view - somehow each page does refill commit cache on first hit

The circumstances in which this will definitely improve last commit cache performance would be dependent on a few factors.

The best improvement would be on pages that do not contain the pathological cases for the caching algorithm, that is long unchanged tree entries. If these are spread throughout the pages this PR won't help and if anything will make it worse.

> somehow each page does refill commit cache on first hit

? I don't understand what you mean. If it's what I think you're implying that kind of thing would need the comments in https://github.com/go-gitea/gitea/pull/16432#issuecomment-886563789 implemented.

---

#16467 is the real significant performance improvement - not this.

**Comment by lafriks** [comment]
Yes I meant that PR but had no time to look at it in details yet. 50 is not that much imho and while it helps performance it hurts usability. I don't think we need to sacrifice usability for large number of repositories to gain performance on very small extreme cases. We need to balance it carefully.

**Comment by 6543** [comment]
> I like the idear but it does not speed up the view - somehow each page does refill commit cache on first hit

solved?

**Comment by delvh** [comment]
> Yes I meant that PR but had no time to look at it in details yet. 50 is not that much imho and while it helps performance it hurts usability. I don't think we need to sacrifice usability for large number of repositories to gain performance on very small number of extreme cases. We need to balance it carefully.

> 50 will make paging even for gitea repo root directory listing...

Well, let me put it this way: Could it be that Gitea **is a large repository**?
On top of that, the files inside gitea that would not be shown would be below the `README` and under normal circumstances, those are the really unimportant files:
Currently, these would be `main.go`, `package-lock.json`, `package.json`, and `webpack.config.js`. Those are not the files you typically look at using the web UI.

In the end, it comes down to how "aesthetic" we want to be:
Of course, we could use the `50` because it is a round number.
But we could just as well use `60`, `65` or even `69` to still get the improved performance.
Given that there is no count and users would have to count it themselves, I would almost prefer `69` as it is the maximum number to still get the performance boost. And with `69`, even Gitea, which definitely counts as a large repo, would still have lots of slots to fill before pagination will be needed.
I have to agree with @zeripath: If you have to scroll above 70 items, it is also not user-friendly.

There is only one potential problem I can foresee: What do you do if the `README` is below that limit? Will it simply not be displayed on the initial page, or what will happen?

File: custom/conf/app.example.ini [hunk 3]
```diff
@@ -1011,6 +1011,9 @@ PATH =
 ;; Number of issues that are displayed on one page
 ;ISSUE_PAGING_NUM = 10
 ;;
+;; Number of entries are displayed on code rendering page
```
Comment (delvh) [inline custom/conf/app.example.ini:1014]:
```suggestion
;; Number of entries displayed on file tree pages
;; Big enough repositories will load significantly slower if the value is >= 70
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 4]
```diff
@@ -169,6 +169,7 @@ The following configuration set `Content-Type: application/vnd.android.package-a
 - `EXPLORE_PAGING_NUM`: **20**: Number of repositories that are shown in one explore page.
 - `ISSUE_PAGING_NUM`: **10**: Number of issues that are shown in one page (for all pages that list issues).
 - `MEMBERS_PAGING_NUM`: **20**: Number of members that are shown in organization members.
+- `DIRECTORY_PAGING_NUM`: **50**: Number of directory entries that are shown in one page.
```
Comment (delvh) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md:172]:
```suggestion
- `DIRECTORY_PAGING_NUM`: **50**: Number of directory entries that are shown in one page. Setting the value >= 70 will slow down loading big enough repositories significantly.
```

File: routers/web/repo/view.go [hunk 5]
```diff
@@ -232,6 +223,35 @@ func renderDirectory(ctx *context.Context, treeLink string) {
...
 
+	// Now limit the number of directory items to the page
+	page := ctx.FormInt("page")
+	if page < 1 {
+		page = 1
+	}
+	pageSize := setting.UI.DirectoryPagingNum
+
+	if pageSize > 1 {
+		pager := context.NewPagination(len(entries), pageSize, page, 5)
```
Comment (delvh) [inline routers/web/repo/view.go:234]: Why the 5?
What does that do?
Is that perhaps the number of page literals to show?

**Review by delvh — COMMENTED** [review]
Regarding my prior comment:
My fear towards the README not being displayed seems not to be a problem.

**Comment by delvh** [comment]
Oh, one thing is unclear to me:
You said that values >= 70 affect performance negatively.
Does that only apply to big enough repos, or does it apply to every repo?

File: modules/setting/setting.go (hunk 1 above)
Comment (6543) [reply modules/setting/setting.go:252]: well this is kinda bike shading and i personally do prefer a larger number ... just no strong opinion ... but it's configurable - so what's the point, smal repos would fix easily

**Comment by 6543** [comment]
@zeripath can you resolve conflict :)

**Pull request closed without merging**